### ISSUE TITLE
youtube-dl issue

### ISSUE THREAD

**Issue by mija** [kepler471/dvembed#3]
I got this issue no matter what:

```
2021/01/21 16:37:02 dvembed bot is now running. Press CTRL-C to exit.
2021/01/21 16:37:17 name https://v.redd.it/8b7qosjl6oc61 [0xc00019c1e0] &[0xc00019c1e0]
2021/01/21 16:37:17 Message by name  contains v.redd.it link: https://v.redd.it/8b7qosjl6oc61
2021/01/21 16:37:17 Message sent on channel: 793927661822935100
2021/01/21 16:37:17 Message segment: https://v.redd.it/8b7qosjl6oc61, is valid URl
2021/01/21 16:37:17 Entering download process >
2021/01/21 16:37:17     ...run youtube-dl...
2021/01/21 16:37:17     Failed youtube-dl process: [youtube-dl -v --output 8b7qosjl6oc61.mp4 --write-info-json --merge-output-format mp4 https://v.redd.it/8b7qosjl6oc61]
2021/01/21 16:37:17 Did not download: exec: "downloads\8b7qosjl6oc61\youtube-dl.exe": file does not exist
```

was unable to solve it until i changed download.go to this:

```
package main

import (
	"log"
	"os"
	"os/exec"
	"path"
	"fmt"
)

type media struct {
	os.FileInfo
	Id   string
	Path string
}

// download handles retrieving the v.redd.it media. Initialises
// a media instance for each URL.
func download(URL string) (*media, error) {
	f := media{
		Id: path.Base(URL),
	}
	dir, _  := os.Getwd()
	cmd := exec.Command(
		fmt.Sprintf("%v/youtube-dl", dir),
		"-v",
		//"--id",              // use id as name
		"--output",
		f.Id+originalExt,
		"--write-info-json", // save file information
		//"--restrict-filenames",
		"--merge-output-format", // Downloading the best available audio and video
		outputFormat,
		URL,
	)
	cmd.Dir = path.Join(dir, f.Id)
	err := os.MkdirAll(cmd.Dir, 0755)
	if err != nil {
		log.Printf("\tError creating sub-directory: %v", err)
		cmd.Dir = dir
	}

	f.Path = path.Join(cmd.Dir, f.Id+originalExt)
	if info, err := os.Stat(f.Path); !os.IsNotExist(err) {
		log.Printf("\tFile at %v exists, will not be downlaoded. Err: %v", f.Path, err)
		f.FileInfo = info
		return &f, nil
	} else {
		log.Print("\t...run youtube-dl...")
		err = cmd.Run()
		if err != nil {
			log.Printf("\tFailed youtube-dl process: %v", cmd.Args)
			return &f, err
		}
	}
	f.FileInfo, err = os.Stat(f.Path)
	if err != nil {
		log.Print("\tError finding downloaded file: ", err)
	}
	// TODO want to return output error from youtube-dl
	return &f, err
}

```

this solved the issue.

**Comment by kepler471** [comment]
Hey! Thanks for taking a look at my project. Out of interest, how did you come accross it?

**Comment by mija** [comment]
v.redd.it is one of those things i just absolutely hate and if we ignore how shit their videoplayer is, how unstable their servers and how it doesnt buffer correctly etc the fact it doesnt get embedded in discord is also a great annoyance.

i googled around hoping to find some sort of solution to it and eventually stumbled into this. there was another similar repo but yours seemed a bit more promising, especially the plans to attempt to hit the filesize limit.

another possible workaround when it comes to filesize is maybe allowing files that were over 8mb to be uploaded to gdrive/onedrive/etc and put in the link to the file which should result in an embed?

**Comment by kepler471** [comment]
That's a nice suggestion for the larger files, I had a similar suggestion by someone else to host and serve the files from the server that is running the bot as well. Both would need setup config, and are definitely possible, but I prefer your idea as the user wouldn't have to maintain the server. 

Firstly I will try the compression that I mention in the comments, as ffmpeg can maintain quality really well, with significant file size reduction (see ffmpeg's `-crf` flag). Seeing as you took a look I'll start working on this again to see if I can get that running :)

Once I figure out a file size that can no longer be reduced to 8MB and still be watchable, I will look at using a service to host the video.

Oh, and I will be pushing a fix that I have been using locally and had forgotten push, to do with a file path existing error.

And there is one other bug, where sometimes the bot will see a reddit link with a still image, and still run it through the process resulting in the bot posting the image as an mp4.

✌️

**Comment by mija** [comment]
Ah I thought that mp4 thing was a quick implementation you did to slap in the image from reddit threads into the discord, but i guess it was just a bug.

noticed it a few times.

**Comment by mija** [comment]
This problem persisted when updating to the new version. I tried pip3 install youtube-dl but get a %PATH% error (2021/02/28 23:22:40 did not download: exec: "youtube-dl": executable file not found in %PATH%) even though the youtube-dl is in %PATH%.

This was with .exe added after youtube-dl in the downloads.go, without .exe added after i got the issue about it only looking for youtube-dl.exe in the download folder of the specific video

so go/src/dvembed/downloads/videoname/downloads/videoname/youtube-dl.exe

which obviously is an issue since it creates a folder for each video and i cant manually move over the youtube-dl.exe into each one.
when .exe is added after youtube-dl in downloads it instead complains about %PATH%.

Ill see if my previous fix for this works again.

**Comment by kepler471** [comment]
Hey, just to double check, did you build from source or use the .exe in the releases? I have not yet compiled the latest release.

**Comment by kepler471** [comment]
Ok I have just set up on windows and it seems to be working , with Python 3.9 installed at C:\Users\%USER%\AppData\Local\Programs\Python\Python39\, and youtube-dl installed with pip was installed at C:\Users\%USER%\AppData\Local\Programs\Python\Python39\Scripts\. Both of these were default locations for the Python installation, and during the installation I let the installer add these directories to PATH.

`pip install youtube-dl` allowed me to use `youtube-dl` or `youtube-dl.exe` from the command line, both work.

I also installed ffmpeg ([download url a build from here](https://ffmpeg.org/download.html#build-windows)). I extracted the folder and renamed it ffmpeg, then placed that folder at /c/Users/%USER%/AppData/Local/Programs/ffmpeg/, and finally, added /c/Users/%USER%/AppData/Local/Programs/ffmpeg/bin/ to the PATH. 

Similarly, both `ffmpeg` and `ffmpeg.exe` worked.

So it should be working. A mistake I made was not to keep ffmpeg in the requirements, ffmpeg is required by youtube-dl to mux the audio and video files it downloads separately, sorry for that confusion!. But other than that, this is working on Windows on my end. Let me know if it still doesnt work with youtube-dl and ffmpeg both properly added to path.
